lib/active_merchant/billing/gateways/simetrik.rb

```diff
@@ -143,49 +139,64 @@ def scrub(transcript)
 
       def add_forward_route(post, options)
         forward_route = {}
-        add_if_has_key(forward_route, options, :trace_id)
+        forward_route[:trace_id] = options[:trace_id] if options[:trace_id]
+
         forward_route[:psp_extra_fields] = options[:psp_extra_fields] || {}
         post[:forward_route] = forward_route
       end
 
       def add_forward_payload(post, money, payment, options)
         forward_payload = {}
-        add_user(forward_payload, options[:user])
-        add_order(forward_payload, money, options[:order])
-        add_payment_method(forward_payload, payment, options[:payment_method])
-        forward_payload[:authentication] = {} unless forward_payload.key?(:authentication)
-        add_three_ds_fields(forward_payload[:authentication], options[:three_ds_fields]) if options[:three_ds_fields]
+        add_user(forward_payload, options[:user]) if options[:user]
+        add_order(forward_payload, money, options[:order]) if options[:order] || money
+        add_payment_method(forward_payload, payment, options[:payment_method]) if options[:payment_method] || payment
```

Comment:
It seems like we are looking for billing address data in `options[:payment_method][:card][:billing_address]`, but the standard Active Merchant API is to provide billing address data in `options[:billing_address]`.

If we simply use `options[:billing_address]` then we could refactor `add_payment_method` to something that is easier to read and maintain, for example:

```
if payment
  add_card(forward_payload, payment)
  add_address(:billing_address, forward_payload[:card], options[:billing_address]) if options[:billing_address]
end
```
Then `add_card` would only be adding payment data, and it would be possible to remove `add_payment_method`.

Reply:
Thanks for the heads-up, following your recommendation, we now take `billing_address` from `options[:billing_address]` instead of `options[:payment_method][:card][:billing_address]`. 

The `add_address('billing_address', card_hash, card_options[:billing_address]) if card_options` line previously found in `add_card` has been removed, and the following block of code has been added to `add_forward_payload`: 

`forward_payload[:payment_method] = {} unless forward_payload[:payment_method]
        forward_payload[:payment_method][:card] = {} unless forward_payload[:payment_method][:card]
        add_address('billing_address', forward_payload[:payment_method][:card], options[:billing_address]) if options[:billing_address]
`